tests/test_dataset_binning.py

```diff
@@ -0,0 +1,127 @@
+import numpy as np
+import pytest
+
+
+@pytest.mark.parametrize("bins", [5, 10, 20, 50, [100.0, 1.0]])
+@pytest.mark.parametrize("inplace", [True, False])
+@pytest.mark.parametrize("return_labels", [True, False])
+@pytest.mark.parametrize("return_edges", [True, False])
+def test_assign_resolution_bins(
+    data_fmodel, bins, inplace, return_labels, return_edges
+):
+    """Test DataSet.assign_resolution_bins() arguments"""
+
+    result = data_fmodel.assign_resolution_bins(
+        bins=bins,
+        inplace=inplace,
+        return_labels=return_labels,
+        return_edges=return_edges,
+    )
+
+    if return_labels and return_edges:
+        result, labels, edges = result
+    elif return_labels:
+        result, labels = result
+    elif return_edges:
+        result, edges = result
+
+    # Test bins
+    assert "bin" in result.columns
+    if isinstance(bins, int):
+        assert len(result["bin"].unique()) == bins
+        assert result.bin.max() == bins - 1
+    else:
+        assert all(result["bin"].to_numpy() == 0)
+
+    # Test inplace
+    if inplace:
+        assert id(result) == id(data_fmodel)
+    else:
+        assert id(result) != id(data_fmodel)
+
+    # Test labels
+    if return_labels:
+        assert isinstance(labels, list)
+        assert all([isinstance(l, str) for l in labels])
+        if isinstance(bins, int):
+            assert len(labels) == bins
+        else:
+            assert len(labels) == len(bins) - 1
+
+    # Test edges
+    if return_edges:
+        assert np.all(np.diff(edges) < 0)
+        assert isinstance(edges, np.ndarray)
+        if isinstance(bins, int):
+            assert len(edges) == bins + 1
+        else:
+            assert all(edges == bins)
+
+
+@pytest.mark.parametrize("bins", [1, 5, 10, 20, 50])
+def test_binedges_equivalence(data_fmodel, bins):
+    """
+    Test DataSet.assign_resolution_bins() generates consistent assignment when
+    provided bin edges from `return_edges=True`
+    """
+
+    expected, edges = data_fmodel.assign_resolution_bins(
+        bins, return_labels=False, return_edges=True
+    )
+
+    result = data_fmodel.assign_resolution_bins(
+        edges, return_labels=False, return_edges=False
+    )
+
+    assert all(expected["bin"] == result["bin"])
+
+
+@pytest.mark.parametrize("reverse", [True, False])
+def test_binedges_order(data_fmodel, reverse):
+    """
+    Test DataSet.assign_resolution_bins() assignments depend on whether bin edges
+    are provided in ascending or descending order
+    """
+
+    expected, edges = data_fmodel.assign_resolution_bins(
+        2, return_labels=False, return_edges=True
+    )
+
+    # Flip bin edges -- all assignments should differ
+    if reverse:
```

Comment:
can this be made to test for correctness? right now it only tests to see the result is not incorrect in one particular way.

Reply:
Sure -- this tests correctness against the assignments made using `assign_resolution_bins(2)`, but I can also check against `dHKL`, explicitly

Reply:
yep -- just updated this to test explicitly against `(1 - bin)` in the reverse case